src/Compilers/CSharp/Portable/Syntax/SyntaxTreeDiagnosticEnumerator.cs

```diff
@@ -11,136 +13,121 @@ namespace Microsoft.CodeAnalysis.CSharp
     /// <summary>
     /// An enumerator for diagnostic lists.
     /// </summary>
-    internal struct SyntaxTreeDiagnosticEnumerator
+    internal static class SyntaxTreeDiagnosticEnumerator
     {
-        private readonly SyntaxTree? _syntaxTree;
-        private NodeIterationStack _stack;
-        private Diagnostic? _current;
-        private int _position;
         private const int DefaultStackCapacity = 8;
 
-        internal SyntaxTreeDiagnosticEnumerator(SyntaxTree syntaxTree, GreenNode? node, int position)
+        public static IEnumerable<Diagnostic> EnumerateDiagnostics(SyntaxTree syntaxTree, GreenNode root, int position)
         {
-            _syntaxTree = null;
-            _current = null;
-            _position = position;
-            if (node != null && node.ContainsDiagnostics)
-            {
-                _syntaxTree = syntaxTree;
-                _stack = new NodeIterationStack(DefaultStackCapacity);
-                _stack.PushNodeOrToken(node);
-            }
-            else
-            {
-                _stack = new NodeIterationStack();
-            }
-        }
+            if (!root.ContainsDiagnostics)
+                yield break;
 
-        /// <summary>
-        /// Moves the enumerator to the next diagnostic instance in the diagnostic list.
-        /// </summary>
-        /// <returns>Returns true if enumerator moved to the next diagnostic, false if the
-        /// enumerator was at the end of the diagnostic list.</returns>
-        public bool MoveNext()
-        {
-            while (_stack.Any())
+            using var stack = new NodeIterationStack(DefaultStackCapacity);
+            stack.PushNodeOrToken(root);
+
+            while (stack.Any())
```

Comment:
can have a simple while loop here, instead of having to store state, return true, and then restart where we were.